environment/environment.go

```diff
@@ -0,0 +1,906 @@
+		},
+	}
+}
+
+// Builds the base kubernetes client
+func kubeClient() (*kubernetes.Clientset, error) {
+	config, err := kubeConfig()
+	if err != nil {
+		return nil, err
+	}
+	return kubernetes.NewForConfig(config)
+}
+
+// Uses default loading rules to load the kubernetest config
+func kubeConfig() (*rest.Config, error) {
+	loadingRules := clientcmd.NewDefaultClientConfigLoadingRules()
+	kubeConfig := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(loadingRules, &clientcmd.ConfigOverrides{})
+	return kubeConfig.ClientConfig()
+}
+
+// Forwards all ports needed to connect to the environment, along with the hardhat port
+func (env *environment) forwardPorts(namespaceName string) ([]string, string, string, error) {
+	kubeClient, err := kubeClient()
+	if err != nil {
+		return nil, "", "", err
+	}
+	kubeConfig, err := kubeConfig()
+	if err != nil {
+		return nil, "", "", err
+	}
+
+	// TODO: Figure out a way to do this for services, the obvious way gives inscrutable errors
+	// https://gianarb.it/blog/programmatically-kube-port-forward-in-go
+	// https://github.com/gianarb/kube-port-forward/issues/3
+	podInterface := kubeClient.CoreV1().Pods(namespaceName)
+	podList, err := podInterface.List(context.Background(), metav1.ListOptions{})
+	if err != nil {
+		return nil, "", "", err
+	}
+
+	log.Info().Msg("Forwarding Ports")
+	forwardedPorts := []string{}
+	hardhatPort := ""
+	adapterPort := ""
+	for _, pod := range podList.Items {
+		// stopCh control the port forwarding lifecycle. When it gets closed the
+		// port forward will terminate
+		stopCh := make(chan struct{}, 1)
+		// readyCh communicate when the port forward is ready to get traffic
+		readyCh := make(chan struct{})
+		// stream is used to tell the port forwarder where to place its output or
+		// where to expect input if needed. For the port forwarding we just need
+		// the output eventually
+		stream := genericclioptions.IOStreams{
+			In:     os.Stdin,
+			Out:    nil, // Changing this the os.Stdout can help with debugging
+			ErrOut: os.Stderr,
+		}
+		// receive the forwarded port when it is generated and ready
+		portCh := make(chan string, 1)
+
+		// managing termination signal from the terminal. As you can see the stopCh
+		// gets closed to gracefully handle its termination.
+		sigs := make(chan os.Signal, 1)
+		signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
+		go func() {
+			<-sigs
+			close(stopCh)
+		}()
+		env.forwardedPortChannels = append(env.forwardedPortChannels, sigs)
+
+		go func() {
+			err := env.forwardPort(portForwardRequest{
+				RestConfig: kubeConfig,
+				Pod:        pod,
+				Streams:    stream,
+				StopCh:     stopCh,
+				ReadyCh:    readyCh,
+				PortCh:     portCh,
+			})
+			if err != nil {
+				log.Err(err).Str("Pod", pod.Name).Msg("Error while forwarding port, tearing down env")
+				_ = env.TearDown()
+				// TODO: Not a fan of this, but otherwise caught in an infinite loop if something goes wrong.
+				panic(err)
```

Comment:
See TODO. The test will hang if something errors here until 10-minute timeout, not sure how else to exit out of the hang.